core/trino-main/src/main/java/io/trino/server/ResultQueryInfo.java

```diff
@@ -0,0 +1,280 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.server;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.errorprone.annotations.Immutable;
+import io.trino.execution.BasicStageInfo;
+import io.trino.execution.ExecutionFailureInfo;
+import io.trino.execution.QueryInfo;
+import io.trino.execution.QueryState;
+import io.trino.spi.ErrorCode;
+import io.trino.spi.QueryId;
+import io.trino.spi.TrinoWarning;
+import io.trino.spi.security.SelectedRole;
+import io.trino.transaction.TransactionId;
+import jakarta.annotation.Nullable;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+@Immutable
+public class ResultQueryInfo
+{
+    private final QueryId queryId;
+    private final QueryState state;
+    private final boolean isScheduled;
+    private final String updateType;
+    private final BasicQueryStats queryStats;
+    private final ErrorCode errorCode;
+    private final Optional<BasicStageInfo> outputStage;
+    private final boolean isFinalQueryInfo;
+    private final ExecutionFailureInfo failureInfo;
+    private final Optional<String> setCatalog;
+    private final Optional<String> setSchema;
+    private final Optional<String> setPath;
+    private final Optional<String> setAuthorizationUser;
+    private final Boolean resetAuthorizationUser;
+    private final Map<String, String> setSessionProperties;
+    private final Set<String> resetSessionProperties;
+    private final Map<String, SelectedRole> setRoles;
+    private final Map<String, String> addedPreparedStatements;
+    private final Set<String> deallocatedPreparedStatements;
+    private final Optional<TransactionId> startedTransactionId;
+    private final boolean clearTransactionId;
+    private final List<TrinoWarning> warnings;
+
+    @JsonCreator
```

Comment:
Why does this class need to be JSON serializable?

Reply:
It is replacement for `QueryInfo` that is also `JSON`-serializable. It is used in query protocol.